src/queues.c

```diff
@@ -350,6 +350,25 @@ void queues_history_pop(void)
         g_queue_insert_sorted(waiting, n, notification_cmp_data, NULL);
 }
 
+/* see queues.h */
+bool queues_history_pop_id(int id)
+{
+        if (g_queue_is_empty(history))
+                return false;
+
+        struct notification *n = queues_get_by_id(id);
+        if (!n)
+                return false;
+
+        if (!g_queue_remove(history, n))
+                return false;
+
+        n->redisplayed = true;
+        n->timeout = settings.sticky_history ? 0 : n->timeout;
+        g_queue_insert_sorted(waiting, n, notification_cmp_data, NULL);
```

Comment:
This can be deduped with `queues_history_pop`. I think the cleanest way is `queues_history_pop_n(notification *n)` and make both functions use it.

Reply:
Sure, I'll take a look.
Also, I noticed that when you replace a notification, the old notification still exists in history. Should we purge the replaced notification from history, or make it so that the history-pop-id function only pops the last version? Right now it pops all version one after the other

Reply:
It's invalid from the spec to replace a notification that has already been closed. So we can leave it as it is.

Reply:
I don't see it mentioned in the spec

Reply:
From `NotificationClosed`:
> The ID specified in the signal is invalidated before the signal is sent and may not be used in any further communications with the server. 

So after that signal no further communication should reference that ID.

Reply:
Oh then this whole feature is invalid according to the spec. I do not see a valid way to achieve this otherwise, other than resending the notification. That kind of defeats the point

Reply:
Hmm, we could do a history-pop with app-id. That would be fine for most cases and easier to use

Reply:
Dunstctl is itself outside the spec either way. The spec only specifies what happens between the client and the server, anything else is extra. So in that space we can do whatever.

Thought I do have doubts about the usefulness of this feature, other than the issue linked I don't how for what else it'd be used for - it's a pretty hacky solution.

Reply:
I think I have to agree with you there. The application could always resend a notification instead. This will probably be a more hacky solution in most cases.

Reply:
Shall I close the PR then?

Reply:
Feel free to close, I'm open to implementing something like this with a valid use-case. One thing I can think of is for scripts in combination with `dunstctl`. But I'm not sure what use popping a notification from history would have.

I guess the real problem in #806 is 'you can't get the icon for a raw icon notification' which is a pretty tough one to solve.

Reply:
> I guess the real problem in #806 is 'you can't get the icon for a raw icon notification' which is a pretty tough one to solve.

It's not really hard to make a working solution for that though. See https://github.com/dunst-project/dunst/issues/804#issuecomment-756697382 for a patch that saves the icon to a file. We could add two rules: one rule to enable icon saving to a file, and a second one to specify where the icon should be saved.

Reply:
That's no really a solution IMO. It adds unnecessary IO writes, and might even introduce race conditions.

A proper solution would be
a) Somehow pass the file contents to the script via bash (I don't think that's possible, bash doesn't like binary content)
b) Somehow request the icon from dunst via dunctl

Any other ideas are appreciated but these are the only paths I see currently.

Reply:
> That's no really a solution IMO. It adds unnecessary IO writes, and might even introduce race conditions.

Race conditions maybe, but you can avoid unneccesary IO by saving the file to a tmpfs (/run and sometimes /tmp as well).

> A proper solution would be
> a) Somehow pass the file contents to the script via bash (I don't think that's possible, bash doesn't like binary content)

Maybe a file descriptor would work? But that's more complicated than it should be.

> b) Somehow request the icon from dunst via dunctl
> 
> Any other ideas are appreciated but these are the only paths I see currently.

I think we need a good way to point to old notifications then, so a better ID system. I'll leave it for now though, but if we come up with more use cases for needing to point to old notifications we can think more about it.